### ISSUE TITLE
Improve company-tng integration

### ISSUE THREAD

**Issue by yyoncho** [emacs-lsp/lsp-mode#1438]
See:

https://www.reddit.com/r/emacs/comments/dp4jcx/lspmode_ccls_wrong_autocomplete/ and especially the YCMD implementation.

https://github.com/company-mode/company-mode/pull/946

The potential solution for how is to introduce a variable that will make lsp-mode prefer the final insert text over the label.

(cc @dgutov @kiennq )

**Comment by dgutov** [comment]
Shouldn't simply setting `lsp-enable-snippet` have this behavior?

Or do you see some value in current behavior as well when snippets are disabled?

**Comment by kiennq** [comment]
No, currently we're using `label` for displaying suggestions in candidate window, and it requires post completion to replace the inserted `label` with `textEdit`. The `label` and `textEdit`'s `newText` can be different, for example: powershell language server.

@yyoncho: Going YCMD way, does that mean we should display `newText` instead of `label` when available? 

@dgutov: I saw that `company-tng` is using `company-prefix` to determine overlay. Can we add additional option for it to look at a certain candidate's property to determine prefix too?
The reason is in `lsp-mode`, we're not getting the prefix up front but rather wait for the response from server to determine correct prefix for current completion.
Currently we just pass `(bounds-of-thing-at-point 'symbol)` as prefix place holder, and later ignore it   and determine our own prefix after getting results from server.

**Comment by yyoncho** [comment]
> @yyoncho: Going YCMD way, does that mean we should display `newText` instead of `label` when available?

The default behaviour stays as it is since it provides richer functionality and it is in line with the reference LSP client implementation VSCode. My proposal is, in addition to that to introduce a mechanism to enable company-tng compatible mode. The mechanism could be enabled either by variable or by inspecting the company backends. 

As a side note, I am working on company-box option to allow following the cursor with completion popup which fixes/avoids the fact that the prefix is not the correct one:
![Peek 2020-02-20 21-09](https://user-images.githubusercontent.com/13259670/74969932-2eddce00-5426-11ea-96d3-46329d4b67ca.gif)

**Comment by dgutov** [comment]
> No, currently we're using label for displaying suggestions in candidate window

I'm just saying you probably don't need a new user option. Just fix the behavior of this one when set to nil.

> Going YCMD way, does that mean we should display newText instead of label when available?

Ideally, I think, you would also display the rest of the string as annotation.

> Can we add additional option for it to look at a certain candidate's property to determine prefix too?

It's doable, but either we do it as a new backend action (and we'll need some good new name), or in an ad-hoc way in the popup frontend which might or might not work with the alternatives, such as company-posframe.

Speaking of company-box, @yyoncho would I be right to guess that you're using macOS?

**Comment by yyoncho** [comment]
> Ideally, I think, you would also display the rest of the string as annotation.

We may try that but it is not nessesary to have the same prefix. E. g. the label for tag might be 'body' but the text to insert can be '<body'. 

> Speaking of company-box, @yyoncho would I be right to guess that you're using macOS?

I am on Linux Mint.

**Comment by dgutov** [comment]
> We may try that but it is not nessesary to have the same prefix. E. g. the label for tag might be 'body' but the text to insert can be '<body'.

So even if the "real prefix" is `<b`, as I'm assuming, the label can still be `body`? That doesn't bode well for the usual lining-up of the popup as we do in company.

> I am on Linux Mint.

Hmm, OK. FWIW, company-box doesn't really work in my GNOME installation.

**Comment by yyoncho** [comment]
> So even if the "real prefix" is <b, as I'm assuming, the label can still be body? That doesn't bode well for the usual lining-up of the popup as we do in company.

Yes, vscode does not align as company-mode do.

> Hmm, OK. FWIW, company-box doesn't really work in my GNOME installation.

What doesn't work on your side? I am using gnome 2 and I did several fixes to make it less noisy.

**Comment by dgutov** [comment]
> What doesn't work on your side? I am using gnome 2 and I did several fixes to make it less noisy.

It looked pretty broken under GNOME 3 (Mutter) when using a GTK3 Emacs build. Tried it again (with some experimental patches aiming to fix problems with company-posframe in the same environment), and it seems like they're hitting similar bugs. So, we'll see.

The GTK2 Emacs build seemed to work okay-ish, however.

**Comment by dgutov** [comment]
> Yes, vscode does not align as company-mode do

Not aligning is one thing, but it looks like the protocol won't allow us to do that, even.

**Comment by Jemoka** [comment]
Hey folks! Is there any forward motion on this? For me, the entire list of arguments is still completing with company-tng.
